src/features/product/contexts/ProductContext.tsx

```diff
@@ -0,0 +1,121 @@
+import {
+  createContext,
+  PropsWithChildren,
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+import { useSearchParams } from "react-router-dom";
+import { Product } from "../../../apis/types/response";
+import useFetch from "../../../shared/hooks/useFetch";
+import { ProductsAPI } from "../apis/ProductsAPI";
+import { CategoryOptionsKey, SortOptionsKey } from "../config/filter";
+
+type ProductParams = {
+  category: CategoryOptionsKey;
+  sortOption: SortOptionsKey;
+};
+
+export interface ProductContextType {
+  refetch: () => Promise<void>;
+  loading: boolean;
+  products: Product[];
+
+  category: {
+    value: CategoryOptionsKey;
+    set: (newCategory: CategoryOptionsKey) => void;
+  };
+
+  sort: {
+    value: SortOptionsKey;
+    set: (newSort: SortOptionsKey) => void;
+  };
+}
+
+export const ProductContext = createContext<ProductContextType | null>(null);
+
+export const ProductProvider = ({ children }: PropsWithChildren) => {
+  const [products, setProducts] = useState<Product[]>([]);
+  const [searchParams, setSearchParams] = useSearchParams();
+
+  const fetchProducts = useCallback(
+    (params: ProductParams) =>
+      ProductsAPI.get(params.category, params.sortOption),
+    []
+  );
+
+  const { data, loading, success, fetchData } = useFetch<
+    Product[],
+    ProductParams
+  >(fetchProducts);
```

Comment:
함수를 통해 타입이 자동으로 유추되도록 하려면 어떻게 해야 좋을까요?

```suggestion
  const { data, loading, success, fetchData } = useFetch(fetchProducts);
```

이렇게 쓰고 싶은거죠 ㅎㅎ